mlflow/projects/__init__.py

```diff
@@ -321,8 +344,12 @@ def run(
 
     if backend == "databricks":
         mlflow.projects.databricks.before_run_validations(mlflow.get_tracking_uri(), backend_config)
-    elif backend == "local" and run_id is not None:
-        backend_config_dict[MLFLOW_LOCAL_BACKEND_RUN_ID_CONFIG] = run_id
+    else:
+        if backend == "local" and run_id is not None:
+            backend_config_dict[MLFLOW_LOCAL_BACKEND_RUN_ID_CONFIG] = run_id
+        if tracking_uri is not None:
+            if _resolve_tracking_uri(tracking_uri) is not None:
+                mlflow.set_tracking_uri(_resolve_tracking_uri(tracking_uri))
```

Comment:
Can we set the tracking URI for the project by propagating it to the project backends: `databricks` & `kubernetes` (https://github.com/thethiny/mlflow/blob/0acb133389f4c89c65ef24f274cde4bb04af1b79/mlflow/projects/__init__.py#L141-L192) as well as pluggable backends inheriting from `AbstractBackend` (https://github.com/thethiny/mlflow/blob/0acb133389f4c89c65ef24f274cde4bb04af1b79/mlflow/projects/backend/abstract_backend.py#L4)?

This way, `mlflow.projects.run()` doesn't have side effects, which is currently the case if we set the global tracking URI via `mlflow.set_tracking_uri()`.

Fortunately, the abstract pluggable backend's `run()` method already accepts a `tracking_uri` parameter, so we should just be able to thread this tracking URI straight through. For Databricks and Kubernetes, we can set the `MLFLOW_TRACKING_URI` environment variable in the resulting cluster environment specification and Kubernetes deployment specification, respectively.

Reply:
I didn't want to get near this because I found that there's a separate handler for databricks that goes through a different pipeline. If I do the change as you suggested the databricks api will be prone to errors.

Reply:
I'm going to address this in a commit on its own as I believe I would require further discussion with you (perhaps different PR?). For now I will handle the other comments.
From what I've seen in the code there are multiple functions that assume their own `tracking_uri` such as https://github.com/thethiny/mlflow/blob/0acb133389f4c89c65ef24f274cde4bb04af1b79/mlflow/projects/databricks.py#L404.
And if I understand you correctly, you want me to update all the functions so that they **receive** and **pass** `tracking_uri` wherever it's applicable so that it can be passed in. Though what I don't understand is that if you want me to pass the `tracking_uri` variable on its own to each and every or if you want me to `mlflow.set_tracking_uri()` in the `run()` function and all functions then will automatically work as intended since they will rely on `mlflow.get_tracking_uri()`.

Example:
Either accept `tracking_uri` in the line shown above and pass it on as `databricks_profile_uri=tracking_uri or tracking.get_tracking_uri()`
or
do the `None` check on tracking uri in the [`run()`](https://github.com/thethiny/mlflow/blob/0acb133389f4c89c65ef24f274cde4bb04af1b79/mlflow/projects/__init__.py#L344) function **only**.